### PR TITLE
Set management call functionality for httpServer

### PR DESCRIPTION
**Pull request by shgayle**
Implemented functionality to allow one call to getManagement for
   all the management commands. Current command set might not be
    complete, so additions might be required.
    Had to rearrange some structures to complete the task, so these
    structures are now in templateMain.go.

    Testing was challenging because of a number of reasons, so careful
    review and testing will be required.

    The testing screen prints are still in the files. It is easier to
    review what is happening with this output. I can do a direct search
    and remove against the template files after review is completed. I
    will also do a spell check, and message review at that time.

    I could not get the error logs to work, wasn't focusing on this, but
    the diagnostic logs are fine.

### CODE REVIEW

File: microservices/ga/workerPool/httpJob.go [hunk 1]
```diff
@@ -5,45 +5,75 @@ import (
...
+//Type: returns the job type
 func (j *httpJob) Type() string {
 	return HTTPJobType
 }
-
+//GetClientID: returns the clientID for the job
+func (j *httpJob) GetClientID() string {
+        return j.ClientID
+}
+//Excount: returns the number of time a continuous job
```
Comment (jscharber) [inline microservices/ga/workerPool/httpJob.go:70]: Needs to be ExCount to keep lint from complaining

File: microservices/ga/workerPool/httpJob.go [hunk 2]
```diff
@@ -58,24 +88,75 @@ func (j *httpJob) Init() error {
 	if j.ClientTimeout == 0 {
 		j.ClientTimeout = ClientTimeout
 	}
+	/*
```
Comment (jscharber) [inline microservices/ga/workerPool/httpJob.go:91]: Do we need to keep this comment?

File: microservices/ga/workerPool/httpJob.go [hunk 3]
```diff
@@ -5,45 +5,75 @@ import (
...
+        JobID         uuid.UUID `json:"job_id"`
+        JobType       string    `json:"job_type"`
+        client        *http.Client
+        ClientTimeout int `json:"client_timeout"`
+        // TODO: FIX to errors or custom errors
+        jobErrors  []string
+        JobURL     *url.URL  `json:"job_url"`
+        Stats      httpStats `json:"stats"`
+        ClientID   string    `json:"client_id"` //Possible business client
+        sent       bool      //Internal
```
Comment (jscharber) [inline microservices/ga/workerPool/httpJob.go]: Internal doesn't say anything // sent job has been written to XX chan

File: microservices/ga/workerPool/httpJob.go [hunk 4]
```diff
@@ -58,24 +88,75 @@ func (j *httpJob) Init() error {
...
+        }
+}
 
-func (j *httpJob) Run() (result Result, err error) {
-		req, err := http.NewRequest("GET", j.JobURL.String(), nil)
+//PausedAsSent is used to pause a continuous job.
+//Job remains on jobList but is not sent.
+func (j *httpJob) PausedAsSent() bool {
+        if j.continuous && !j.sent {
+                //After completion a continous job is
```
Comment (jscharber) [inline microservices/ga/workerPool/httpJob.go]: I think you mean it is, not is is.

Isn't a job created by reference?  If so why do I need to mark it as not sent?

File: microservices/ga/workerPool/httpJob_test.go [hunk 5]
```diff
@@ -23,13 +23,14 @@ func JobTestMain() {
 	go startServer()
 
 	// Wait for server to start
-	time.Sleep(1 * time.Second)
+	time.Sleep(5 * time.Second)
```
Comment (jscharber) [inline microservices/ga/workerPool/httpJob_test.go:26]: Do we really need five seconds, I was going to reduce this?

File: microservices/ga/workerPool/httpJob_test.go [hunk 6]
```diff
@@ -64,8 +66,9 @@ func TestHTTPJob(t *testing.T) {
 }
 
 func TestHTTPJob_Metrics(t *testing.T) {
+	fmt.Println("TestHTTPJob_Metrics")
 	j := newJob("http://" + ServerString + EPString)
-	_, e := j.Run()
+	_, e := j.Run(&Mwg)
```
Comment (jscharber) [inline microservices/ga/workerPool/httpJob_test.go:71]: This change requires the interface to change, need to discuss it

File: microservices/ga/workerPool/httpScheduler.go [hunk 7]
```diff
@@ -152,6 +191,34 @@ func (s *httpScheduler) GetScheduledJobs() ([]byte, error) {
 	return jb, nil
 }
 
+// StopContinuousJob returns the status for a single job status change request
+func (s *httpScheduler) StopContinuousJob(UUID string) (httpStatusCode int, jsonBlob []byte, err error) {
+        msg := ""
+
+        for ji := range s.jobList {
```
Comment (jscharber) [inline microservices/ga/workerPool/httpScheduler.go:198]: range returns two values?

File: microservices/ga/workerPool/httpScheduler.go [hunk 8]
```diff
@@ -352,8 +468,51 @@ func (s *httpScheduler) SetChannels(j chan Job, r chan Result, b chan bool, i ch
...
+
+        if value < 0 {
+                value = value * -1
+        }
+        if value > defaultMaxJobCount {
+                value = defaultMaxJobCount
+       }
+
+        old := value
+        s.mutex.Lock()
```
Comment (jscharber) [inline microservices/ga/workerPool/httpScheduler.go:497]: I liked where you added the defer, I'd keep that consistent throughout

File: microservices/ga/workerPool/httpScheduler.go [hunk 9]
```diff
@@ -352,8 +468,51 @@ func (s *httpScheduler) SetChannels(j chan Job, r chan Result, b chan bool, i ch
...
+        rmsg := fmt.Sprintf("{\" Status\": \"%s changed from %d to %d\"}",
+                trackMaxJobs, old, value)
+        return http.StatusOK, []byte(rmsg), nil
+}
+
 // Process methods
-func (s *httpScheduler) Init() error {
+func (s *httpScheduler) Init(mo *managementCommands) error {
+
+	var funcPtr interface{}
```
Comment (jscharber) [inline microservices/ga/workerPool/httpScheduler.go:511]: This isn't a point?

File: microservices/ga/workerPool/httpScheduler.go [hunk 10]
```diff
@@ -362,60 +521,135 @@ func (s *httpScheduler) Init() error {
...
 	}
 
 	return nil
 }
 
-func (s *httpScheduler) Run() error {
-	go s.RunScheduler()
-	go s.RunResultsReader()
+func (s *httpScheduler) Run(mWg *sync.WaitGroup) error {
+	mWg.Add(2)
```
Comment (jscharber) [inline microservices/ga/workerPool/httpScheduler.go:571]: All these goroutines need a ctx

File: microservices/ga/workerPool/templateDispatcher.go [hunk 11]
```diff
@@ -47,141 +61,209 @@ const (
...
-// swagger:response managementGetResponse
-type managementGetResponse struct {
-	// The error message
-	// in: body
-
-	// Commands is a list of valid commands that can be executed
-	Commands []mgtCommand `json:"commands"`
-	// Fields is a list of fields that can be changed
-	Fields []string `json:"fields"`
+func (d *dispatcher) stopScheduler() (httpStatusCode int, jsonb []byte, err error) {
```
Comment (jscharber) [inline microservices/ga/workerPool/templateDispatcher.go:64]: Looks like we lost the swagger documentation

File: microservices/ga/workerPool/templateDispatcher.go [hunk 12]
```diff
@@ -47,141 +61,209 @@ const (
...
-		JsonBody string `json:"json_body"`
-	} `json:"body"`
+func (d *dispatcher) stopWorkers() (httpStatusCode int, jsonb []byte, err error) {
+	d.metrics.mutex.Lock()
+        if !ChannelsReady {
+                d.metrics.mutex.Unlock()
+                msg := fmt.Sprintf("{\"status\": \"can't stop workers now\"}")
+                return http.StatusForbidden, []byte(msg), nil
+        }
+        fmt.Println("in stoping wrokers")
```
Comment (jscharber) [inline microservices/ga/workerPool/templateDispatcher.go:97]: We should make these type of debug messages conditional on a debug level

File: microservices/ga/workerPool/template_test.go [hunk 13]
```diff
@@ -30,130 +32,176 @@ const (
...
+	JobTestMain()
 	defer os.Exit(m.Run())
+	defer testExit()
+	Mwg.Wait()
 }
 
 func testExit() {
-	data := "{\"command\": \"shutdown_now\", \"field\": \"\", \"field_value\": 0}"
-	req, _ := http.NewRequest("PUT", ManagementURL, strings.NewReader(data))
+        mr.CommandName = "shutdown_now"
```
Comment (jscharber) [inline microservices/ga/workerPool/template_test.go:84]: Let's talk about naming conventions.

File: microservices/ga/workerPool/httpJob.go [hunk 14]
```diff
@@ -5,45 +5,75 @@ import (
...
+	"github.com/google/uuid"
 	"io/ioutil"
 	"net/http"
 	"net/url"
+	"strconv"
+	"sync"
 	"time"
-
-	"github.com/google/uuid"
 )
```
Comment (jscharber) [inline microservices/ga/workerPool/httpJob.go:16]: There is a lot of code changes.  I need to check out this branch and play around with it.

**Comment by shgayle** [comment]
Re:
We should make these type of debug messages conditional on a debug level

I was anticipating  clean up after your review. All fmt.Println statements
will be removed, unless you really want conditional testing statements left
in for debugging.

-Solomon

**Comment by jscharber** [comment]
Moved to closed

**Pull request closed without merging**